common/util/container_proxy.h

```diff
@@ -0,0 +1,454 @@
+//     `const ContainerType& underlying_container() const`:
+//     Returns reference to the wrapped container.
+//
+// - Optional:
+//   - `void ElementsInserted(iterator first, iterator last)`:
+//     Called when new elements were inserted. `first` and `last` (exclusive)
+//     refer to inserted elements.
+//   - `void ElementsBeingRemoved(iterator first, iterator last)`:
+//     Called just before removing elements from `first` to `last` (exclusive).
+//   - `void ElementsBeingReplaced()`:
+//     Called when all elements are going to be replaced (due to assignment).
+//   - `void ElementsWereReplaced()`:
+//     Called just after all elements were replaced.
+//
+// Usage
+// -----
+//
+// 1. Subclass:
+//
+//        using ContainerType = std::vector<int>
+//        class MyProxy: private ContainerProxyBase<MyProxy, ContainerType> {
+//          using Base = ContainerProxyBase<MyProxy, ContainerType>;
+//          friend Base;
+//          // ...
+//        };
+//
+// 2. Implement mandatory interface:
+//
+//        private:
+//          ContainerType& underlying_container() {
+//            return container_;
+//          }
+//          const ContainerType& underlying_container() const {
+//            return container_;
+//          }
+//        // ...
+//        ContainerType container_; // not a part of the interface
+//
+// 3. Inherit public interface (all or only some members):
+//
+//        public:
+//          using typename Base::container_type;
+//          using typename Base::value_type;
+//          using typename Base::reference;
+//          using typename Base::const_reference;
+//          // ...
+//          using Base::begin;
+//          using Base::end;
+//          // ...
+//
+//    Note: inheriting (or implementing) `void swap(MyProxy&)` will also provide
+//    standalone function `void swap(MyProxy&, MyProxy&)`.
+//
+// 4. Implement optional interface:
+//
+//        private:
+//          void ElementsInserted(iterator first, iterator last) {
+//            /* ... */
+//          }
+//          void ElementsBeingRemoved(iterator first, iterator last) {
+//            /* ... */
+//          }
+//          void ElementsBeingReplaced() { /* ... */ }
+//          void ElementsWereReplaced() { /* ... */ }
+//
+// 5. Optional: implement operator=(const MyProxy&) and operator=(MyProxy&&)
+//    (base class' `operator=` works only with `ContainerType`):
+//
+//        public:
+//          MyProxy& operator=(const MyProxy& other) {
+//            *this = other.underlying_container();
+//            return *this;
+//          }
+//          MyProxy& operator=(MyProxy&& other) {
+//            // You probably want to notify `other` about invalidation
+//            // of its **whole container** (not just elements).
+//            *this = std::move(other.underlying_container());
+//            return *this;
+//          }
+//
+template <class DerivedType, class ContainerType>
+class ContainerProxyBase {
```

Comment:
The name is not sufficiently conveying what this class does.
I think it should be possible to see that this provides way to observe change, so maybe something along the lines of `ObservableContainerDelegatorBase` ... even though it is a mouth-full.
Maybe you have better ideas.

We can also defer the decision to somewhere in the near future once we have a good name, so that we don't stall the current effort.

Reply:
I prefer to defer this decision. I've added a TODO comment in the code.

Ideas/notes for future reference:
- Something like "ContainerObserverBase". The derived class is the one that observes a container. Delegation is an option. I assume that since the base class is a CRTP base/template it shouldn't be viewed as a separate entity that delegates something.
- It would be useful to convey that the class not only observes, but also provides container-like interface.